doc/notes/cost-model-representations/cost-model-representations.md

```diff
@@ -0,0 +1,216 @@
+other), and the memory cost is 1 (required to store the boolean result of the
+comparison).
+
+    "lessThanEqualsByteString": {
+        "cpu": {
+            "arguments": {
+                "intercept": 197145,
+                "slope": 156
+            },
+            "type": "min_size"
+        },
+        "memory": {
+            "arguments": 1,
+            "type": "constant_cost"
+        }
+    }
+
+We refer to the numbers in this structure as the _cost model parameters_.
+
+
+### Internal Haskell representations
+There are a number of different Haskell representations of these structures in the Plutus codebase,
+and it can be difficult to work out which ones are used where.
+
+#### The main representation
+
+The main internal representation of the cost model is given by the
+[BuiltinCostModelBase](https://github.com/input-output-hk/plutus/blob/e773e58ea0e4a8088fed0ea5f934a7c413caa5b3/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/BuiltinCostModel.hs#L70)
+type, which is a record type with one field for each built-in function. The type
+is parametric over a type `f`.  In reality we instantiate `f` to be the
+`CostingFun` type defined
+[here](https://github.com/input-output-hk/plutus/blob/e773e58ea0e4a8088fed0ea5f934a7c413caa5b3/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/CostingFun/Core.hs#L75)
+to obtain the [BuiltinCostModel](https://github.com/input-output-hk/plutus/blob/e773e58ea0e4a8088fed0ea5f934a7c413caa5b3/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/BuiltinCostModel.hs#L57) type.
+
+When the UPLC evaluator is compiled the contents of `builtinCostModel.json` are
+read in and converted into a BuiltinCostModel object called
+[defaultBuiltinCostModel](https://github.com/input-output-hk/plutus/blob/e773e58ea0e4a8088fed0ea5f934a7c413caa5b3/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExBudgetingDefaults.hs#L36).
+The costs for the different types of CEK steps (all of which currently have the
+same cost) are defined in
+[cekMachineCosts.json](https://github.com/input-output-hk/plutus/blob/master/plutus-core/cost-model/data/cekMachineCosts.json)
+which is compiled into a CekMachineCosts object called [defaultCekMachineCosts](https://github.com/input-output-hk/plutus/blob/b321575d9266b3358b9e728d064fc0bee4f355d7/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExBudgetingDefaults.hs#L65),
+and this is bundled together with the builtin costs in [defaultCekCostModel](https://github.com/input-output-hk/plutus/blob/b321575d9266b3358b9e728d064fc0bee4f355d7/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExBudgetingDefaults.hs#L69) (of
+type [CostModel CekMachineCosts BuiltinCostModel](https://github.com/input-output-hk/plutus/blob/e773e58ea0e4a8088fed0ea5f934a7c413caa5b3/plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/MachineParameters.hs#L28)).
+
+#### Flattened cost model entries
+The cost model is sometimes flattened into a list of pairs of (key, value)
```

Comment:
Maybe something like: "While the nested structure is convenient for representing the structure of the parameters, for interoperating with other sources of parameter values (such as the ledger) it is convenient to have a flat representation of the parameters."?

Reply:
This should all be in the haddock for `CostModelParams` or `CostModelInterface`

Reply:
Can we put it there, but keep this too?  I wrote this down because I was finding it hard to see how all the pieces fit together.

Reply:
Sure, I just think if we're going to the trouble of writing this we should put as much of it into the hopefully more discoverable haddocks.

Reply:
> > have a flat representation of the parameters."?
> 
> I don't like "flat representation" because it reminds me flat, the serialization format

I see what you mean.  Would "flattened" be OK?  I can't think of a better way of putting it.

Reply:
Yes, flattened better